proseco/guide.py

```diff
@@ -153,13 +160,49 @@ def run_search_stages(self):
                 self.log(f'Quitting after stage {idx - 1} with {already_selected} stars', level=1)
                 break
         self.log('Done with search stages')
-        selected = cand_guides[cand_guides['stage'] != -1]
-        selected.sort(['stage', 'mag'])
-        if len(selected) >= self.n_guide:
-            return selected[0:n_guide]
+        stage_cands = cand_guides[cand_guides['stage'] != -1]
+        stage_cands.sort(['stage', 'mag'])
+        selected = self.select_catalog(stage_cands[0:n_guide + GUIDE.surplus_stars])
         if len(selected) < self.n_guide:
             self.log(f'Could not find {self.n_guide} candidates after all search stages')
-            return selected
+        return selected
+
+    def select_catalog(self, stage_cands):
+        """
+        For the candidates selected at any stage (stage-assigned candidates?) select the
+        first combination that satisfies the most number of extra tests.
+        Here the extra tests are just the 3 checks for guide stars that are
+        too tightly clustered.
+        """
+        self.log(f'Selecting catalog from {len(stage_cands)} stage-selected stars')
+        # Set some arrays to save "check status" and combination identifiers so
+        # we only have to search through the candidate combinations ones.  The plan is
+        # if no combination satisfies all the checks, lower the standard (required success
+        # on n checks) and repeat, but do this just searching through the previously
+        # determined status.
+        saved_status = []
+        saved_idxs = []
+
+        # I should come back to this and see if the "min" is needed or if the combinations
+        # code runs fine even if we have fewer-than-expected stage_cands to start
+        choose_n = min(len(stage_cands), self.n_guide)
+
+        for comb in combinations(range(len(stage_cands)), choose_n):
+            cluster_status = run_comb_checks(stage_cands[[comb]])
+            saved_idxs.append(comb)
+            saved_status.append(cluster_status)
+            if len(saved_idxs) > 1:
+                self.required_optimization = True
+            if np.all(cluster_status):
+                self.log(f'Selected stars after trying {len(saved_idxs)} combinations')
+                return stage_cands[[comb]]
+        for n_success in [2, 1, 0]:
```

Comment:
For a first go, instead of deciding which of the checks was most important or most likely to fail, I just set the code to allow for 2 of 3, 1 of 3, and then no success when running through the checks.